File: engine/src/types/state.ts
// ── State Types ─────────────────────────────────────────────

export type StateScope = "page" | "app";

export interface StateDefinition {
  key: string;
  scope: StateScope;
  initial: unknown;
}

export interface PageState {
  scope: "page";
  data: Record<string, unknown>;
}

export interface AppState {
  scope: "app";
  data: Record<string, unknown>;
}

export type State = PageState | AppState;

File: engine-dart/lib/src/types/state.dart
/// Declares initial state for a page.
class StateDefinition {
  final String key;
  final String scope; // "page" | "app"
  final dynamic initial;

  const StateDefinition({
    required this.key,
    required this.scope,
    required this.initial,
  });

  factory StateDefinition.fromJson(Map<String, dynamic> json) {
    return StateDefinition(
      key: json['key'] as String,
      scope: json['scope'] as String,
      initial: json['initial'],
    );
  }

  Map<String, dynamic> toJson() => {
        'key': key,
        'scope': scope,
        'initial': initial,
      };
}
